src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -317,31 +320,37 @@
      * Constructor, defaults to {@link #GITHUB_URL} as the end-point, and anonymous access, does not default any
      * {@link SCMSourceTrait} behaviours.
      *
-     * @param repoOwner the repository owner.
-     * @param repository the repository name.
-     * @param repositoryUrl HTML URL for the repository. If specified, takes precedence over repoOwner and repository.
-     * @param configuredByUrl Whether to use repositoryUrl or repoOwner/repository for configuration.
+     * This was choosen as the constructor to allow users to set only the values they need and
+     * not have to pass extra null or empty values. However, this source cannot be used until
+     * either "repositoryUrl" or "repoOwner"/"repository" are set.
+     * If both combinations are set, repositoryUrl will be used.
+     *
+     * See {@link #bindResolve} for further discussion of the reasons for this design.
+     *
      * @throws IllegalArgumentException if repositoryUrl is specified but invalid.
      * @since 2.2.0
      */
-    // configuredByUrl is used to decide which radioBlock in the UI the user had selected when they submitted the form.
     @DataBoundConstructor
-    public GitHubSCMSource(String repoOwner, String repository, String repositoryUrl, boolean configuredByUrl) {
-        if (!configuredByUrl) {
-            this.apiUri = GITHUB_URL;
-            this.repoOwner = repoOwner;
-            this.repository = repository;
-            this.repositoryUrl = null;
-        } else {
-            GitHubRepositoryInfo info = GitHubRepositoryInfo.forRepositoryUrl(repositoryUrl);
-            this.apiUri = info.getApiUri();
-            this.repoOwner = info.getRepoOwner();
-            this.repository = info.getRepository();
-            this.repositoryUrl = info.getRepositoryUrl();
-        }
+    public GitHubSCMSource() {
         pullRequestMetadataCache = new ConcurrentHashMap<>();
         pullRequestContributorCache = new ConcurrentHashMap<>();
-        this.traits = new ArrayList<>();
+        traits = new ArrayList<>();
+        repoOwner = "";
+        repository = "";
```

Comment:
Pretty weird, might mask some bugs. Personally, I would rather see an NPE caused by the instance being in an illegal state than have the instance silently do the wrong thing without any way for us to diagnose the issue.

Maybe assign the fields to null and add `SuppressFBWarnings` mentioning that the setters will make the fields non-null in practice, or change the annotation to `@CheckForNull` (but that's going to require a bunch of changes in the methods here). Same thing for the `defaultIfBlank` calls in the setters.

Reply:
@dwnusbaum  Community PR to do that conversion are fine with me.  😄 

As to the assigning blank, you and @darxriggs both commented on it. And you're right, but ... this exactly the point of this PR - DataBinding with a complex set of parameters basically demands that we construct the instance in non-ready-to-run state. On the other hand, for any valid use of of an instance these values will not be null. On the **other** other hand, we don't make them null by default because we want to be able to load/save instances that are in a non-ready-to-run state so that we can modify/fix them to be ready-to-run.